polynote-kernel/src/main/scala/polynote/kernel/dependency/DependencyFetcher.scala

```diff
@@ -208,6 +185,76 @@ class CoursierFetcher extends DependencyFetcher[IO] {
     run(resolution.process, 0)
   }
 
+  def splitDependencies(deps: List[DependencyConfigs]): (List[URI], List[DependencyConfigs]) = {
+    val (urlList, dependencies) = deps.flatMap { dep =>
+      dep.toList.collect {
+        case (k, v) =>
+          val (dependencies, urls) = v.map { s =>
+            Either.fromTry {
+              Try(new URL(s)) // use URL because its parser is strict, even though we actually want URI
+                .map(_.toURI)
+                // we might still know how to handle the protocol even if URL doesn't
+                .recoverWith {
+                  case t: MalformedURLException if t.getMessage.contains("unknown protocol") => Try(new URI(s))
+                }
+            }.leftMap(_ => s) // on the left, strings are dependency coordinates
+          }.separate
+
+          (urls, TinyMap(k -> TinyList(dependencies)))
+      }
+    }.unzip
+    (urlList.flatten, dependencies)
+  }
+
+  def fetchUrls(uris: List[URI], statusUpdates: Publish[IO, KernelStatusUpdate], chunkSize: Int = 8192)(implicit ctx: ExecutionContext): List[(String, IO[File])] = {
+
+    def withProgress(s: fs2.Stream[IO, Byte], uri: String, fileSize: Long): fs2.Stream[IO, Byte] = {
+      val size = if (fileSize > 0) fileSize else 300 * 1024 * 1024 // this is roughly the size of the largest uber jars we have
+
+      // is there a better way to do this? seems more convoluted than necessary but I couldn't quite get there...
```

Comment:
@jeremyrsmith is there a nicer way to implement this progress thingie?

Reply:
I dunno... looks pretty nice to me actually 😁 

I'd say, compare it to what you'd do without fs2, just using `publish1().unsafeRunSync()` to publish the status, and if that looks cleaner (ignoring the `unsafe`) then go with it.

That said, I think considering what you're accomplishing (getting a file download with progress updates sent through a websocket to the client) that this is pretty darn clean already! Cleaner even than the equivalent thing in `CoursierFetcher`, FWIW.

Reply:
ok, I guess the thing I didn't like is mainly that its weird that undoing `Stream.chunks` needs to be done with a `flatMap(Stream.chunk)` but I guess that's more of an FS2 issue. It was a bit strange that `unchunk` didn't undo `chunk`, so I was a bit suspicious that there might be something I was missing in their API.